zebra-chain/src/keys/sprout.rs

```diff
@@ -1,14 +1,115 @@
-use std::{
-    fmt,
-    io::{self},
-};
+//! Sprout key types
+//!
+//! "The receiving key sk_enc, the incoming viewing key ivk = (apk,
+//! sk_enc), and the shielded payment address addr_pk = (a_pk, pk_enc) are
+//! derived from a_sk, as described in [‘Sprout Key Components’][ps]
+//!
+//! [ps]: https://zips.z.cash/protocol/protocol.pdf#sproutkeycomponents
+
+use byteorder::{ByteOrder, LittleEndian};
 
 #[cfg(test)]
 use proptest::{array, collection::vec, prelude::*};
 #[cfg(test)]
 use proptest_derive::Arbitrary;
 
-use crate::serialization::{SerializationError, ZcashDeserialize, ZcashSerialize};
+use sha2::compress256;
+
+/// Our root secret key of the Sprout key derivation tree.
+///
+/// All other Sprout key types derive from the SpendingKey value.
+/// Actually 252 bits.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct SpendingKey(pub [u8; 32]);
+
+/// Derived from a _SpendingKey_.
+pub type ReceivingKey = x25519_dalek::StaticSecret;
+
+impl From<SpendingKey> for ReceivingKey {
+    /// For this invocation of SHA256Compress as the PRF, t=0, which
+    /// is populated by default in an empty block of all zeros to
+    /// start.
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#sproutkeycomponents
+    /// https://zips.z.cash/protocol/protocol.pdf#concreteprfs
+    fn from(spending_key: SpendingKey) -> ReceivingKey {
+        let mut state = [0u32; 8];
+        let mut block = [0u8; 64]; // Thus, t = 0
+
+        block[0..32].copy_from_slice(&spending_key.0[..]);
+        block[0] |= 0b1100_0000;
+
+        compress256(&mut state, &block);
```

Comment:
nit: this might be slightly clearer if it were `sha2::compress256` rather than having to backtrack to the import statement above